src/lib/form-field/form-field.ts

```diff
@@ -220,24 +227,40 @@ export class MatFormField implements AfterViewInit, AfterContentInit, AfterConte
     return ngControl && (ngControl as any)[prop];
   }
 
-  /** Whether the form field has a placeholder. */
   _hasPlaceholder() {
     return !!(this._control.placeholder || this._placeholderChild);
   }
 
+  _hasLabel() {
+    return !!this._labelChild;
+  }
+
+  _shouldLabelFloat() {
+    return this._canLabelFloat && (this._control.shouldLabelFloat ||
+        this._control.shouldPlaceholderFloat || this._shouldAlwaysFloat);
+  }
+
+  _hideControlPlaceholder() {
+    return !this._hasLabel() || !this._shouldLabelFloat();
```

Comment:
Maybe the `_canLabelFloat` from above should be factored into this logic instead?

Reply:
The `_shouldLabelFloat` call here already factors in `_canLabelFloat`.